src/main/java/picard/sam/CheckDuplicateMarking.java

```diff
@@ -0,0 +1,165 @@
+            if (reader.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.queryname) {
+                log.info("Input file isn't queryname sorted. Sorting into temp space.");
+
+                final Path[] tmpPaths = TMP_DIR.stream()
+                        .map(File::toPath)
+                        .toArray(Path[]::new);
+
+                final ProgressLogger sortProgress = new ProgressLogger(log, (int) 1e6, "Read into sorter");
+                final SortingCollection<SAMRecord> alignmentSorter = SortingCollection.newInstance(SAMRecord.class,
+                        new BAMRecordCodec(reader.getFileHeader()),
+                        new SAMRecordQueryNameComparator(),
+                        MAX_RECORDS_IN_RAM,
+                        tmpPaths);
+
+                for (final SAMRecord rec : reader) {
+                    alignmentSorter.add(rec);
+                    sortProgress.record(rec);
+                }
+                samRecordIterable = alignmentSorter;
+            } else {
+                samRecordIterable = reader;
+            }
+
+            checkDuplicateMarkingsInIterable(samRecordIterable);
+            log.info("Found " + numBadRecords + " bad records.");
+            return numBadRecords > 0 ? 1:0;
+
+        } catch (IOException e) {
+            throw new PicardException("Error while reading input file " + INPUT, e);
+        }
+    }
+
+    private boolean duplicateMarkingGood(final SAMRecord rec) {
+        if (!rec.getReadName().equals(currentReadName)) {
+            currentReadName = rec.getReadName();
+            currentReadDuplicateMarked = rec.getDuplicateReadFlag();
+            return true;
+        } else if (rec.getDuplicateReadFlag() != currentReadDuplicateMarked) {
+            if (numBadRecords <= NUM_WARNINGS) {
+                log.warn(() -> "Reads with queryname " + currentReadName + " have different duplicate flags (at " +
+                        rec.getContig() + ":" + rec.getStart() + ")");
+            }
+            if (numBadRecords == NUM_WARNINGS) {
+                log.warn("Further warnings will be suppressed.");
+            }
+
+            numBadRecords++;
+            return false;
+        }
+        return true;
+    }
+
+
+    private void checkDuplicateMarkingsInIterable(final Iterable<SAMRecord> iterable) throws IOException {
+        try (PrintWriter writer =  OUTPUT==null?
+                new PrintWriter(NullOutputStream.NULL_OUTPUT_STREAM):
+                new PrintWriter(new FileWriter(OUTPUT))) {
+
+            for (final SAMRecord rec : iterable) {
+                if (MODE != Mode.ALL && rec.isSecondaryOrSupplementary()) {
+                    continue;
+                }
+
+                if (MODE == Mode.PRIMARY_MAPPED_ONLY && rec.getReadUnmappedFlag()) {
+                    continue;
+                }
+
+                if (MODE == Mode.PRIMARY_PROPER_PAIR_ONLY && !rec.getProperPairFlag()) {
+                    continue;
+                }
+
```

Comment:
No case in code to handle `PRIMARY_ONLY`, is that supported as it's the inverse of the rest?

Reply:
it's a bit confusing, but notice that the first `if` is `if NOT EQUALS Mode.ALL`